core/embed/trezorhal/optiga/optiga.c

```diff
@@ -465,108 +530,150 @@ int optiga_pin_set(OPTIGA_UI_PROGRESS ui_progress,
+  if (res != OPTIGA_SUCCESS) {
+    return res;
+  }
+
+  // Generate the key for the HMAC-SHA256 PIN stretching step.
+  uint8_t pin_hmac[OPTIGA_PIN_SECRET_SIZE] = {0};
+  res = optiga_get_random(pin_hmac, sizeof(pin_hmac));
   if (res != OPTIGA_SUCCESS) {
-    memzero(stretched_pin, sizeof(stretched_pin));
     return res;
   }
+  random_xor(pin_hmac, sizeof(pin_hmac));
 
-  // Authorise using OID_PIN_SECRET so that we can write to OID_PIN_COUNTER and
-  // OID_STRETCHED_PIN.
+  // Authorise using OID_PIN_SECRET so that we can write to OID_STRETCHED_PIN
+  // and OID_STRETCHED_PIN_CTR.
   res = optiga_set_auto_state(OPTIGA_OID_SESSION_CTX, OID_PIN_SECRET,
-                              out_secret, OPTIGA_PIN_SECRET_SIZE);
+                              pin_secret, sizeof(pin_secret));
   if (res != OPTIGA_SUCCESS) {
-    memzero(stretched_pin, sizeof(stretched_pin));
     return res;
   }
 
-  // Set the stretched PIN.
-  res = optiga_set_data_object(OID_STRETCHED_PIN, false, stretched_pin,
-                               sizeof(stretched_pin));
-  memzero(stretched_pin, sizeof(stretched_pin));
+  // Process the stretched PIN using a one-way function before using it in the
+  // operation that will be executed in Optiga during verification. This ensures
+  // that in the unlikely case of an attacker recording communication between
+  // the MCU and Optiga, they will not gain knowledge of the stretched PIN.
+  uint8_t digest[OPTIGA_PIN_SECRET_SIZE] = {0};
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
+
+  // Compute the operation that will be executed in Optiga during verification.
+  uint8_t hmac_buffer[ENCRYPT_SYM_PREFIX_SIZE + OPTIGA_PIN_SECRET_SIZE] = {
+      0x61, 0x00, 0x20};
+  hmac_sha256(pin_hmac, sizeof(pin_hmac), digest, sizeof(digest),
+              &hmac_buffer[ENCRYPT_SYM_PREFIX_SIZE]);
+
+  // Stretch the PIN with the result.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, hmac_buffer,
+              sizeof(hmac_buffer), stretched_pin);
+  memzero(hmac_buffer, sizeof(hmac_buffer));
+
+  // Process the stretched PIN using a one-way function before sending it to the
+  // Optiga. This ensures that in the unlikely case of an attacker recording
+  // communication between the MCU and Optiga, they will not gain knowledge of
+  // the stretched PIN.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
+
+  // Store the digest of the stretched PIN in OID_STRETCHED_PIN.
+  res =
+      optiga_set_data_object(OID_STRETCHED_PIN, false, digest, sizeof(digest));
   if (res != OPTIGA_SUCCESS) {
-    optiga_clear_auto_state(OID_PIN_SECRET);
     return res;
   }
 
-  // Initialize the PIN counter.
-  res = optiga_set_data_object(OID_PIN_COUNTER, false, COUNTER_RESET,
+  // Initialize the counter which limits the guesses at OID_STRETCHED_PIN so
+  // that we can authorise using OID_STRETCHED_PIN.
+  res = optiga_set_data_object(OID_STRETCHED_PIN_CTR, false, COUNTER_RESET,
```

Comment:
Wouldn't it be enough to set the counter to one higher value instead of resetting it later again?

Reply:
That would require a refactor, since we work with the number of depleted attempts, not the number of remaining attempts.

Reply:
We discussed it in person, but we did not reach a consensus.